cEP-0015.md

```diff
@@ -0,0 +1,93 @@
+
+|Metadata|                                                           |
+|--------|-----------------------------------------------------------|
+|cEP     |15                                                         |
+|Version |1.0                                                        |
+|Title   |corobo Enhancement (security, tests and configurability)   |
+|Authors |Nitanshu Vashistha <nitanshu.vzard@gmail.com>              |
+|Status  |Proposed                                                   |
+|Type    |Process                                                    |
+
+Abstract
+--------
+
+This cEP describes the details of enhancement of [corobo](https://github.com/coala/corobo) in terms of security, tests, configurability and
+the new plugins that are to be added to corobo as part of the [GSoC Project](https://summerofcode.withgoogle.com/projects/#6603667076022272).
+
+## Security
+
+### Why improve security?
+
+Security has been one of the major concerns due to some past experiences.
+We want to provide access controls based on user’s GitHub team memberships.
+
+### How to improve security?
+
+1. Make all LabHub commands except invite require being a member of the organization.
+2. Add ability to ban from all gitter rooms at once.
+3. Remove auto-invite and give invite permissions to developers.
+4. Prevent newcomers from self-assigning themselves and give developers permission
+   to assign newcomers.
+5. Prevent corobo being used to spam a room.
+6. Force newcomers to finish one issue first.
+7. Require newcomers to find a newcomer issue to work on before they are invited to
+   join the organization.
+8. Add ability for maintainers to run corobo in safe mode in emergency situations.
+
+## Tests
+
+### Why improve existing tests and infrastructure?
+
+The default TestBase provided by errbot is not enough for testing plugins like LabHub,
+which required intensive mocking. One of the examples of this case is issues like [this](https://github.com/errbotio/errbot/issues/1086)
+
+
+### How to improve existing tests and infrastructure?
+
+This will involve making changes upstream in [Errbot](https://github.com/errbotio/errbot/) and extend the
+existing testing infrastructure to implement better testing for plugins like LabHub.
+
+## Configurability
+
+### Why make corobo more configurable?
+
+corobo has a potential to be used by other organizations for similar tasks like onboarding
+and automation. Currently, it is not configurable and many plugins are still very
+coala specific. Making it more configurable will allow other organizations to adapt corobo
+to cater their needs.
+
+### How to make corobo more configurable?
+
+Currently, the LabHub plugins are coala specific as they are meant to work for a specific
+team name like Maintainers, developers, and newcomers which are coala specific and might be
+different for other organizations.
+
+List of Plugins are:
+
+1. invite
+2. assign
+3. unassign
+4. mark wip|pending
```

Comment:
Actually all four of this commands are part of LabHub plugin, so not a `List of plugins` per se. No other plugins that are coala specific, are you sure?

Reply:
`coatils` for one is coala specific and will remain coala specific, maybe move such specific plugins to a different directory and prevent them from loading for non-coala setup in `Dockerfile`, maybe?

Reply:
Other coala specific plugins are:
1. answers  -- we can maybe make this coala independent and make it a generic docs based answering plugin, will require extensive refactoring
2. coala_lowercase_c  - coala specific only, can't adapt
3. explain -- hardcoded coala explanations
4. searchdocs -- ?

Reply:
@meetmangukiya we can create a separate plugin from `contrib_stats <username>` cmd of `coatils` to show stats of contributions made to the organization. This one can be useful for other organizations?

Reply:
And regarding `corobo explain <arg>` , I think this one is already generalized.
One can simply just add a template in dir 'plugins/templates/explain/` to add new explainations of his choice?

Reply:
maybe have two sub-directories `explain/generic` and `explain/<org name>` , so this repo can be a depot for explains of other orgs.

Reply:
For searchDocs plugin, currently API_DOCS and USER_DOCS URLs are constants for coala, we can make them generic by making the orgs setup their URL/URLs in config file.